topdown/jsonschema.go

```diff
@@ -0,0 +1,109 @@
+		if !json.Valid([]byte(x)) {
+			return nil, errors.New("invalid JSON string")
+		}
+		loader = gojsonschema.NewStringLoader(string(x))
+	case ast.Object:
+		// In case of object serialize it to JSON representation.
+		var data interface{}
+		data, err = ast.JSON(value)
+		if err != nil {
+			return nil, err
+		}
+		loader = gojsonschema.NewGoLoader(data)
+	default:
+		// Any other cases will produce an error.
+		return nil, errors.New("wrong type, expected string or object")
+	}
+
+	return loader, nil
+}
+
+func newResultTerm(valid bool, data *ast.Term) *ast.Term {
+	return ast.ArrayTerm(ast.BooleanTerm(valid), data)
+}
+
+// builtinJSONSchemaValidate accepts 1 argument which can be string or object and checks if it is valid JSON schema.
+// Returns array [false, <string>] with error string at index 1, or [true, ""] with empty string at index 1 otherwise.
+func builtinJSONSchemaValidate(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Take first argument and make JSON Loader from it.
+	loader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return iter(newResultTerm(false, ast.StringTerm("jsonschema: "+err.Error())))
+	}
+
+	// Check that schema is correct and parses without errors.
+	if _, err = gojsonschema.NewSchema(loader); err != nil {
+		return iter(newResultTerm(false, ast.StringTerm("jsonschema: "+err.Error())))
+	}
+
+	return iter(newResultTerm(true, ast.NullTerm()))
+}
+
+// builtinJSONMatchSchema accepts 2 arguments both can be string or object and verifies if the document matches the JSON schema.
+// Returns an array where first element is a boolean indicating a successful match, and the second is an array of errors that is empty on success and populated on failure.
+// In case of internal error returns empty array.
+func builtinJSONMatchSchema(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Take first argument and make JSON Loader from it.
+	// This is a JSON document made from Rego JSON string or object.
+	documentLoader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return err
+	}
+
+	// Take second argument and make JSON Loader from it.
+	// This is a JSON schema made from Rego JSON string or object.
+	schemaLoader, err := astValueToJSONSchemaLoader(operands[1].Value)
+	if err != nil {
+		return err
+	}
+
+	// Use schema to validate document.
+	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
+	if err != nil {
+		return err
+	}
+
+	// In case of validation errors produce Rego array of objects to describe the errors.
+	arr := ast.NewArray()
+	for _, re := range result.Errors() {
+		o := ast.NewObject(
+			[...]*ast.Term{ast.StringTerm("error"), ast.StringTerm(re.String())},
+			[...]*ast.Term{ast.StringTerm("type"), ast.StringTerm(re.Type())},
+			[...]*ast.Term{ast.StringTerm("field"), ast.StringTerm(re.Field())},
+			[...]*ast.Term{ast.StringTerm("desc"), ast.StringTerm(re.Description())},
```

Comment:
Do we need both `re.String()` and `re.Description()`? I mean do they contain different enough that can help in debug etc. or is one them enough ? https://sourcegraph.com/github.com/open-policy-agent/opa@7b8fa09ff8fb3d447abeefa5d257fbc00e4a2ee6/-/blob/internal/gojsonschema/result.go?L159

Reply:
Looks like the string representation contains `field` and `description`, which we include in the error object; but it also includes `context` and `value`. Unless the formatting of the error string makes it much more readable, perhaps we could skip including it, and instead include `context` and `value`(?).
Although, I don't see harm in including it, as the formatting of the string might make the data more human readable.

Reply:
Sounds good 👍